wellcomeml/ml/bert_classifier.py

```diff
@@ -1,236 +1,191 @@
-                with self.textcat.model.use_params(optimizer.averages):
-                    Y_test_pred = self.predict(X_test)
-                    p, r, f1, _ = precision_recall_fscore_support(
-                        Y_test, Y_test_pred, average="micro"
-                    )
-                loss = losses["trf_textcat"]
-                self.losses.append(loss)
-
-                epoch_seconds = time.time() - start_epoch
-                speed = nb_examples / epoch_seconds
-                logger.info(
-                    "{0:5d}\t{1:.3f}\t{2:.3f}\t{3:.3f}\t{4:.3f}\t{5:.2f}".format(
-                        i, loss, p, r, f1, speed
-                    )
-                )
-        return self
 
-    def partial_fit(self, X, Y, classes=None):
-        if not hasattr(self, "unique_labels"):
-            self.unique_labels = [str(i) for i in range(Y.shape[1])]
-            self._init_nlp()
-            self.losses = []
-
-        texts = X
-
-        train_tags = self._label_binarizer_inverse_transform(Y)
-        train_cats = [
-            {label: label in tags for label in self.unique_labels}
-            for tags in train_tags
-        ]
-        annotations = [{"cats": cats} for cats in train_cats]
-
-        other_pipes = [pipe for pipe in self.nlp.pipe_names if "trf" not in pipe]
-        with self.nlp.disable_pipes(*other_pipes):  # only train textcat
-            if not hasattr(self, "optimizer"):
-                self.optimizer = self.nlp.resume_training()
-                self.optimizer.alpha = self.learning_rate
-
-            losses = {}
-            self.nlp.update(
-                texts, annotations, sgd=self.optimizer, drop=self.dropout, losses=losses
-            )
-            self.losses.append(losses["trf_textcat"])
+        train_data = self._prepare_data(X_train, Y_train, shuffle=True, repeat=True)
+        val_data = self._prepare_data(X_test, Y_test)
+
+        steps_per_epoch = math.ceil(len(X_train) / self.batch_size)
+        val_steps_per_epoch = math.ceil(len(X_test) / self.batch_size)
+
+        optimizer = tf.keras.optimizers.Adam(learning_rate=self.learning_rate)
+        if self.multilabel:
+            loss = tf.keras.losses.BinaryCrossentropy(from_logits=True)
+        else:
+            loss = tf.keras.losses.CategoricalCrossentropy(from_logits=True)
+        #  TODO: Fix error when adding metrics
+        #  metrics = [tf.keras.metrics.Precision(), tf.keras.metrics.Recall()]
+        self.model.compile(optimizer=optimizer, loss=loss)
+
+        history = self.model.fit(
+            train_data, validation_data=val_data, epochs=self.epochs,
+            steps_per_epoch=steps_per_epoch, validation_steps=val_steps_per_epoch)
+        self.losses = history.history["loss"]
         return self
 
+    def _predict_proba(self, X):
+        dataset = self._prepare_data(X)
+        out = self.model.predict(dataset)[0]
+        if self.multilabel:
+            return tf.nn.sigmoid(out).numpy()
```

Comment:
I see. I do ```tf.keras.activations.sigmoid/softmax```, but I think the functions just call one another on TensorFlow.

Reply:
agree